tensorflow_datasets/core/features/rle_feature/rle/np_impl.py

```diff
@@ -0,0 +1,276 @@
+"""Numpy encode/decode/utility implementations."""
```

Comment:
What's the motivation to having both a NumPy implementation and a TF implementation? It seems that with Eager mode enabled, the usage of the TF version would be nearly the same.

Reply:
... ... ... I'm not actually sure. Historically I've always developed numpy versions then tried to optimize for tensorflow, and eager mode still isn't part of my standard work flow.

That said, how would one use `tf` implementations in the encoding phase and maintain compatibility with tensorflow 1.x? My understanding is eager mode has to be enabled at the beginning of a script and there's no guarantee when encoding will be run - or is there some way to turn it on and off dynamically?

Also, without an `np.repeat` tensorflow equivalent, there are some ops I'm not equipped to deal with in tensorflow (never written a custom op, and while I'd be keen to get stuck in and learn how this rabbit hole is deep enough as it is).

I can definitely appreciate that numpy encoding implementations and tf decoding implementations are sufficient for most use cases. Adding numpy decoding isn't much trouble (easier than tensorflow until eager mode becomes the norm) and given I already had it I thought I'd throw it in. I accept that's not the most compelling argument though :).

Reply:
I think if we can get the TF implementation to work, we can leave out the numpy version. One fewer thing to have around (maintenance, testing, etc.)

Reply:
I've removed -almost- all duplication. Numpy implementations are used for encoding, tf for decoding. The only duplication is a couple of 1-liners (`rle_length` and `brle_length`).

Is there a simple way to using tf ops in the encoding stage? Once eager mode is default I can see it being straight forward, but until then I'm not sure how to handle it. I have very little experience with eager mode much in 1.x .

If there's an easy way around that and the `np` implementations are still an issue I can refactor them to `tf` variants, but I don't think there'll be much performance improvement if any (and that will be limited to the serialization stage), possibly at the cost of readability. If API stability is the main issue I could make the np versions private and make tf wrappers that can be optimized later if deemed necessary.

Reply:
You could use TF in encode if you wanted to with:

```
with tf.Graph().as_default():
  tensor = ...TF ops...
  np_val = tfds.as_numpy(tensor)
```

If the NP and TF implementations here are actually disjoint, then it's fine to leave them in NP. If there's duplication, rm the NP stuff.

Reply:
... mabe I'm misunderstanding, but this results in ~700x slower code on my fairly ordinary machine for a simple example (equivalent to `brle_length`).

```python
import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds
import timeit

max_val = 1000
size = 1000


def f_tf():
    x = np.random.randint(max_val, size=size)
    with tf.Graph().as_default():
        return tfds.as_numpy(tf.reduce_sum(x))


def f_np():
    x = np.random.randint(max_val, size=size)
    return np.sum(x)


print(timeit.timeit(f_tf, number=1000))     # ~7 sec
print(timeit.timeit(f_np, number=1000))     # ~ 0.01 sec
````

That's using `tf 1.13`. That can be sped up a bit by reusing the same graph/session
```python
graph = tf.Graph()
with graph.as_default():
    x_pl = tf.placeholder(shape=(size,), dtype=tf.int64)
    out = tf.reduce_sum(x_pl)

with tf.Session(graph=graph) as sess:
    def f_sess():
        x = np.random.randint(max_val, size=size)
        return sess.run(out, feed_dict={x_pl: x})

    print(timeit.timeit(f_sess, number=1000))  # 1.5 sec
```

but that's still a heap slower, and requires additional infrastructure to check if `tf.executing_eagerly()`. Even using eager execution things don't get much better.
```python
import numpy as np
import tensorflow as tf
import timeit
tf.compat.v1.enable_eager_execution()

max_val = 1000
size = 1000

def f_tf_eager():
    x = np.random.randint(max_val, size=size)
    return tf.reduce_sum(x)

print(timeit.timeit(f_tf_eager, number=1000))  # ~1.2 sec
```

The story is largely the same for 2.0 preview.

Throwing some `with tf.device('/cpu:0)`s in there plus eager mode or graph/session reuse can get it down to about `0.25`, and some burn-in runs gets you to `0.13`... but we're still taking a 13x speed decrease (~65x when not in graph mode).

 That seems like a high price to pay just to avoid duplication of a couple of 1-line functions. If the only use by `tfds` is during the encoding stage (where the pipeline isn't generally dominated by tensorflow ops) should the `tf` versions be dropped to avoid the duplication? tf implementations _could_ be useful to users, though it's fairly unlikely. In these cases (where `brle_length` is literally just a `tf.reduce_sum` rather than a `np.sum` and `rle_length` isn't much trickier) users could probably work it out themselves, but I'm encountering other situations (not related to this PR) during encoding where numpy implementations are significantly different to tf implementations (mostly in cases with scatter updates or dynamic looping).